File: frontend/src/types/surveillance.ts
/** Phase 3 surveillance API shapes (backend/app/schemas/surveillance.py). */

export type RiskLevelName = 'LOW' | 'MODERATE' | 'ELEVATED' | 'HIGH' | 'CRITICAL';

export interface VesselRiskSummary {
  vessel_id: number;
  score: number;
  level: RiskLevelName;
  computed_at: string;
  top_factor: string | null;
}

export interface RiskFactor {
  type: string;
  score: number;
  explanation: string;
  event_ids: number[];
}

export interface Evidence {
  id: number;
  event_id: number | null;
  evidence_type: string;
  factor_type: string;
  strength: number;
  confidence: number;
  source: string;
  timestamp: string;
  latitude: number;
  longitude: number;
  description: string;
}

export interface VesselRisk {
  vessel_id: number;
  risk_score_id: number;
  score: number;
  level: RiskLevelName;
  factors: RiskFactor[];
  computed_at: string;
  evidence: Evidence[];
}

export interface AISPosition {
  id: number;
  vessel_id: number;
  timestamp: string;
  latitude: number;
  longitude: number;
  speed_over_ground: number | null;
  course_over_ground: number | null;
  heading: number | null;
  navigation_status: string | null;
  source: string;
}

export interface DarkPeriod {
  id: number;
  vessel_id: number;
  start_time: string;
  end_time: string | null;
  duration_seconds: number;
  last_latitude: number;
  last_longitude: number;
  reappearance_latitude: number | null;
  reappearance_longitude: number | null;
  estimated_distance_km: number | null;
  severity: string;
}

export interface SurveillanceEvent {
  id: number;
  event_type: string;
  vessel_id: number;
  other_vessel_id: number | null;
  zone_kind: string | null;
  zone_id: number | null;
  zone_name: string | null;
  timestamp: string;
  latitude: number;
  longitude: number;
  score: number | null;
  confidence: number;
  source: string;
  payload: Record<string, any>;
}

export interface GeoPolygon {
  type: string;
  coordinates: number[][][];
}

export interface FishingZone {
  id: number;
  name: string;
  zone_type: 'AUTHORIZED_FISHING' | 'RESTRICTED_FISHING' | 'SEASONAL_FISHING' | 'NO_FISHING' | string;
  jurisdiction: string | null;
  geometry: GeoPolygon;
}

export interface ProtectedArea {
  id: number;
  name: string;
  protection_level: string;
  authority: string | null;
  rules: string | null;
  geometry: GeoPolygon;
}

export interface InvestigationCase {
  id: number;
  vessel_id: number;
  risk_score: number;
  risk_level: RiskLevelName;
  status: 'OPEN' | 'UNDER_REVIEW' | 'ESCALATED' | 'RESOLVED' | 'DISMISSED';
  assigned_to: string | null;
  summary: string;
  dismissed_reason: string | null;
  agent_summary: string | null;
  created_at: string;
  updated_at: string;
}

export interface InvestigationCaseDetail extends InvestigationCase {
  vessel: { id: number; name: string; vessel_type: string; mmsi: string | null; flag: string | null };
  evidence_snapshot: Record<string, any>[];
  audit_log: { timestamp: string; action: string; actor: string; role?: string | null; note: string | null }[];
}

export interface ScenarioInfo {
  name: string;
  vessel_count: number;
  duration_minutes: number;
}

export interface SimulationRunResult {
  scenario: string;
  start_time: string;
  end_time: string;
  vessels_created: number;
  positions_added: number;
  dark_periods_added: number;
  events_added: number;
  cases_opened: number;
  reset: boolean;
}

/** Live messages on /ws/telemetry. */
export interface LiveSurveillanceEvent {
  event_type: string;
  source: string;
  timestamp: string;
  vessel_id: number | null;
  payload: Record<string, any>;
}

export interface ReplayStep {
  scenario: string;
  sim_time: string;
  step: number;
  total_steps: number;
  progress: number;
  vessels: { mmsi: string; vessel_id: number | null; latitude: number; longitude: number; speed_knots: number; course: number }[];
}

export interface ReplayStart {
  status: string;
  scenario: string;
  steps: number;
  step_seconds: number;
  start_time: string;
  end_time: string;
  dark_windows: { mmsi: string; name: string; start: number; end: number }[];
}

export interface BehaviorProfile {
  source: 'HISTORICAL' | 'LEARNED';
  point_count: number;
  hours_observed: number;
  average_speed: number;
  speed_stddev: number;
  course_change_rate_deg_per_hour: number;
  gap_count: number;
  common_cells: string[];
  window_start: string | null;
  window_end: string | null;
  last_updated: string;
}

export interface BehaviorDeviation {
  score: number;
  speed_z: number;
  baseline_speed: number;
  recent_speed: number;
  turning_ratio: number;
  new_gaps: number;
  explanation: string;
  timestamp: string;
}

export interface VesselBaseline {
  vessel_id: number;
  profile: BehaviorProfile;
  deviation: BehaviorDeviation | null;
}

File: backend/app/schemas/surveillance.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class AISPositionResponse(BaseModel):
    id: int
    vessel_id: int
    timestamp: datetime
    latitude: float
    longitude: float
    speed_over_ground: Optional[float] = None
    course_over_ground: Optional[float] = None
    heading: Optional[float] = None
    navigation_status: Optional[str] = None
    source: str

    model_config = ConfigDict(from_attributes=True)


class DarkPeriodResponse(BaseModel):
    id: int
    vessel_id: int
    start_time: datetime
    end_time: Optional[datetime] = None
    duration_seconds: float
    last_latitude: float
    last_longitude: float
    reappearance_latitude: Optional[float] = None
    reappearance_longitude: Optional[float] = None
    estimated_distance_km: Optional[float] = None
    severity: str

    model_config = ConfigDict(from_attributes=True)


class SimulationRunRequest(BaseModel):
    scenario: str = Field(..., min_length=1)
    start_time: Optional[datetime] = None
    reset: bool = True  # wipe this scenario's earlier data so replays are clean


class SimulationRunResponse(BaseModel):
    scenario: str
    start_time: datetime
    end_time: datetime
    vessels_created: int
    positions_added: int
    dark_periods_added: int
    events_added: int
    cases_opened: int
    reset: bool


class ScenarioInfo(BaseModel):
    name: str
    vessel_count: int
    duration_minutes: int


class SurveillanceEventResponse(BaseModel):
    id: int
    event_type: str
    vessel_id: int
    other_vessel_id: Optional[int] = None
    zone_kind: Optional[str] = None
    zone_id: Optional[int] = None
    zone_name: Optional[str] = None
    timestamp: datetime
    latitude: float
    longitude: float
    score: Optional[float] = None
    confidence: float
    source: str
    payload: dict

    model_config = ConfigDict(from_attributes=True)


class FishingZoneResponse(BaseModel):
    id: int
    name: str
    zone_type: str
    jurisdiction: Optional[str] = None
    allowed_vessel_types: Optional[str] = None
    active_from: Optional[datetime] = None
    active_until: Optional[datetime] = None
    geometry: dict

    model_config = ConfigDict(from_attributes=True)


class ProtectedAreaResponse(BaseModel):
    id: int
    name: str
    protection_level: str
    authority: Optional[str] = None
    rules: Optional[str] = None
    geometry: dict

    model_config = ConfigDict(from_attributes=True)


class RiskFactorResponse(BaseModel):
    type: str
    score: int
    explanation: str
    event_ids: List[int] = []


class EvidenceResponse(BaseModel):
    id: int
    event_id: Optional[int] = None
    evidence_type: str
    factor_type: str
    strength: float
    confidence: float
    source: str
    timestamp: datetime
    latitude: float
    longitude: float
    description: str

    model_config = ConfigDict(from_attributes=True)


class VesselRiskResponse(BaseModel):
    vessel_id: int
    risk_score_id: int
    score: float
    level: str
    factors: List[RiskFactorResponse]
    computed_at: datetime
    evidence: List[EvidenceResponse] = []


class VesselRiskSummary(BaseModel):
    vessel_id: int
    score: float
    level: str
    computed_at: datetime
    top_factor: Optional[str] = None


class CaseVessel(BaseModel):
    id: int
    name: str
    vessel_type: str
    mmsi: Optional[str] = None
    flag: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class InvestigationCaseResponse(BaseModel):
    id: int
    vessel_id: int
    risk_score: float
    risk_level: str
    status: str
    assigned_to: Optional[str] = None
    summary: str
    dismissed_reason: Optional[str] = None
    agent_summary: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InvestigationCaseDetail(InvestigationCaseResponse):
    vessel: CaseVessel
    evidence_snapshot: List[dict]
    audit_log: List[dict]


class AssignRequest(BaseModel):
    assignee: str = Field(..., min_length=1)
    actor: str = Field("operator", min_length=1)


class CaseActionRequest(BaseModel):
    actor: str = Field("operator", min_length=1)
    note: Optional[str] = None


class DismissRequest(CaseActionRequest):
    reason: str = Field(..., min_length=1)


class ReplayRequest(BaseModel):
    scenario: str = Field(..., min_length=1)
    step_seconds: float = Field(0.5, ge=0.0, le=10.0)  # wall-clock pause between scripted reports
    run_analysis: bool = True  # ingest, detect and score before animating


class BehaviorProfileResponse(BaseModel):
    source: str
    point_count: int
    hours_observed: float
    average_speed: float
    speed_stddev: float
    course_change_rate_deg_per_hour: float
    gap_count: int
    common_cells: List[str]
    window_start: Optional[datetime] = None
    window_end: Optional[datetime] = None
    last_updated: datetime

    model_config = ConfigDict(from_attributes=True)


class BehaviorDeviationResponse(BaseModel):
    score: int
    speed_z: float
    baseline_speed: float
    recent_speed: float
    turning_ratio: float
    new_gaps: int
    explanation: str
    timestamp: datetime


class VesselBaselineResponse(BaseModel):
    vessel_id: int
    profile: BehaviorProfileResponse
    deviation: Optional[BehaviorDeviationResponse] = None


class ReplayDarkWindow(BaseModel):
    mmsi: str
    name: str
    start: float  # fraction of the scenario span
    end: float


class ReplayStartResponse(BaseModel):
    status: str
    scenario: str
    steps: int
    step_seconds: float
    start_time: datetime
    end_time: datetime
    dark_windows: List[ReplayDarkWindow] = []
